### PR TITLE
Use Platform* Graphics APIs to match with the rest of the platforms

### LINKED ISSUE #402: Documentation - load image resource doesn't compile on Windows
### Description

https://docs.microsoft.com/en-us/dotnet/maui/user-interface/graphics/paint#create-an-imagepaint-object
This section shows how to draw an image on a canvas.
Loading the image doesn't work _**(does not compile)**_ on Windows (it only works on Android (I have tried Windows and Android)) and there is not any mention of this in the article and there is not any mention of how to load the image on Windows.
Isn't MAUI supposed to be cross-platform?
Isn't there any nice and clean code to load the image cross-platform?

I can see there that build action "EmbeddedResource" must be used instead of "MauiImage" which is default.
I assume that EmbeddedResource creates something like a byte[] resource while MauiImage creates a whole set of images which have specific resolution for each platform etc.

Do I have to put a duplicate image into the resources just so that the same image can be used as a image in ImageButton and at the same time drawn on a canvas (the first one requires MauiImage and the second one EmbeddedResource)?

Thank you

Edit:
- added more info:

`error CS0234: The type or namespace name 'Platform' does not exist in the namespace 'Microsoft.Maui.Graphics' (are you missing an assembly reference?)
`

image

image

### Steps to Reproduce

Try the example on Windows

### Version with bug

Release Candidate 1 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

Windows 10

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #457: Help me! The type or namespace name 'Platform' does not exist in the namespace 'Microsoft.Maui.Graphics'
The type or namespace name 'Platform' does not exist in the namespace 'Microsoft.Maui.Graphics'

pls i cannot find Platform Class when build MAUI windows app 

code :

public void Draw(ICanvas canvas, RectF dirtyRect)
    {
        // Drawing code goes here

        IImage image;
        Assembly assembly = GetType().GetTypeInfo().Assembly;
        using (Stream stream = assembly.GetManifestResourceStream("MemeCreator.Resources.Images.dotnet_bot.svg"))
        {
            image = PlatformImage.FromStream(stream); // error here

        }

        if (image != null)
        {
            IImage newImage = image.Resize(100, 60, ResizeMode.Stretch, true);
            canvas.DrawImage(newImage, 10, 10, newImage.Width, newImage.Height);
        }
    }

Screenshot 2022-06-22 213120

### LINKED ISSUE #722: For Windows, How do I load an image resource and draw it on a GraphicsView?
I simply want to draw an image resource onto a `GraphicsView` (on Microsoft's Windows platform).  This seems like a fairly basic task.  After all, it's a "Multi-plAtform UI" right? :)

So, I fired up Google, found this documentation, and followed the sample code.  After failing to succeed, I went back to the documentation and noticed I missed this warning:

"Warning: The PlatformImage type isn't supported on Windows."

Any chance the page can be updated to provide something more useful? Perhaps, something like "On the Windows platform..." do this instead?  I'm really trying to like MAUI, but it simply doesn't seem ready for prime time yet.

I hope these early experiences don't scare other people away from it.  For my part, I'm going to wait 6 months and try again.  I was super-excited about its promise, but less so now,

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0ca86947-204f-9c45-223d-724fb59b4938
* Version Independent ID: 0ca86947-204f-9c45-223d-724fb59b4938
* Content: [Draw graphical objects - .NET MAUI](https://docs.microsoft.com/en-us/dotnet/maui/user-interface/graphics/draw)
* Content Source: [docs/user-interface/graphics/draw.md](https://github.com/dotnet/docs-maui/blob/main/docs/user-interface/graphics/draw.md)
* Product: **dotnet-mobile**
* Technology: **dotnet-maui**
* GitHub Login: @davidbritch
* Microsoft Alias: **dabritch**

### LINKED ISSUE #820: No way to load image on Windows
Example states that "The PlatformImage type isn't supported on Windows." and doesn't mention any way to load an image in windows.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0ca86947-204f-9c45-223d-724fb59b4938
* Version Independent ID: 0ca86947-204f-9c45-223d-724fb59b4938
* Content: [Draw graphical objects - .NET MAUI](https://docs.microsoft.com/en-us/dotnet/maui/user-interface/graphics/draw)
* Content Source: [docs/user-interface/graphics/draw.md](https://github.com/dotnet/docs-maui/blob/main/docs/user-interface/graphics/draw.md)
* Product: **dotnet-mobile**
* Technology: **dotnet-maui**
* GitHub Login: @davidbritch
* Microsoft Alias: **dabritch**

### LINKED ISSUE #1122: MAUI images not cross-platform on Windows?
On this page, in the "Draw an Image" section, there is a notation that `PlatformImage` is not supported on Windows.  It seems crazy to have a "**Multi-Platform** Application UI", from Microsoft, that supports all platforms **except** Microsoft Windows.
 
I've asked this before, but I was hoping, with the release of .NET 7, that this might have changed. However, it appears it is still an issue.

Am I really forced to write platform-specific code for something as simple as drawing an image on a canvas? Is there even a way to do this on Windows?

Any idea if/when there will be a mechanism for drawing images on a canvas that is **truly** multi-platform?

It seems this should be slightly embarrassing for Microsoft...that an example in their documentation has to explicitly exclude their own operating system.

For further examples of the havoc this causes, I direct you to the following page: [.NET MAUI / Graphics / Images](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/graphics/images).

The many, many notes/warnings about the incompatibilities inherent in MAUI image handling reinforce the impression that MAUI is "not ready for prime time". I honestly hope this impression, and the delays resolving it, won't negatively impact its future.

Thanks in advance for your time.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 0ca86947-204f-9c45-223d-724fb59b4938
* Version Independent ID: 0ca86947-204f-9c45-223d-724fb59b4938
* Content: [Draw graphical objects - .NET MAUI](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/graphics/draw?view=net-maui-7.0)
* Content Source: [docs/user-interface/graphics/draw.md](https://github.com/dotnet/docs-maui/blob/main/docs/user-interface/graphics/draw.md)
* Product: **dotnet-mobile**
* Technology: **dotnet-maui**
* GitHub Login: @davidbritch
* Microsoft Alias: **dabritch**

### LINKED ISSUE #8509: Microsoft.Maui.Graphics.Platform error
### Description

I am trying to display photos in a NetMaui app by using the GraphicsView control.
This worked properly until the latest release of Maui, when the following error appeared:
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0234	The type or namespace name 'Platform' does not exist in the namespace 'Microsoft.Maui.Graphics' (are you missing an assembly reference?)	TestMauiGraphics (net6.0-windows10.0.19041.0)	E:\Projects\NetMAUI\TestMauiGraphics\TestMauiGraphics\MainPage.xaml.cs	35	Active
The error is not shown as marked on the source by VS2022 17,3.0 Preview 2.0
I have created a sample project to show the error at https://github.com/pgmariotti/TestMauiGraphics
Please tell me what I am doing wrong...

### Steps to Reproduce

Just compile the program

### Version with bug

6.0.400 (current)

### Last version that worked well

6.0.312

### Affected platforms

Windows

### Affected platform versions

Windows 11 SDK 10.0.22000.0

### Did you find any workaround?

No

### Relevant log output

_No response_

### LINKED ISSUE #12476: PlatformStringSizeService issues
### Description

I started the code for this project in maui back in .net 6.0 and version 1.0 was published under .net 6.0 https://play.google.com/store/apps/details?id=org.seekingyhwh.calendar . Version 1.1 has been pushed with .net 7.0. Along the way as maui has been improved and changed various problems have been introduced. And one thing I did to get it to compile is to remove all other platforms than android. Right now I'm trying to get it to compile and run for windows. But the class PlatformStringSizeService does not exist for windows even though exists in Microsoft.Maui.Graphics for android. I have been using it because in the past it seemed using a constant number on different android simulators and physical devices along with MS Windows did not always work. I tried to create a project to reproduce the issue, but it's not even compiling for android, and I don't know what I have done differently. https://github.com/sretz/Reproduce.PlatformStringSizeService

### Steps to Reproduce

1. Download https://github.com/sretz/Reproduce.PlatformStringSizeService
2. Change to Windows Machine
3. will not compile

### Link to public reproduction project repository

https://github.com/sretz/Reproduce.PlatformStringSizeService

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows

### Affected platform versions

windows10.0.19041.0

### Did you find any workaround?

No

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change

**⚠️ Breaking Changes ⚠️**

Basic idea of this PR is to align Windows graphics APIs naming with the rest of the platforms. 

The main change is to make the `ShapeHandler` use a new `MauiShapeView` (like all the other platforms) and not the actual platform view.

### Breaking Changes 

 - New `MauiShapeView` type in Core
 - `IShapeViewHandler` and `ShapeViewHandler` now use `MauiShapeView` instead of `W2DGraphicsView`

We can potentially reduce the number of changes by making Core reference Win2D again, but maybe this is going to cause issues. We can also potentially try make W2D version inherit from Platform versions so all the old code mostly works, but now we start off with obsolete things. But that may be required.

### Issues Fixed

 - Fixes #12476 
 - Fixes #8509
 - Fixes https://github.com/dotnet/Microsoft.Maui.Graphics/issues/457
 - Fixes https://github.com/dotnet/Microsoft.Maui.Graphics/issues/402
 - Fixes https://github.com/dotnet/docs-maui/issues/722
 - Fixes https://github.com/dotnet/docs-maui/issues/820
 - Fixes https://github.com/dotnet/docs-maui/issues/1122

### Alternatives
 - https://github.com/dotnet/maui/pull/12578
 - https://github.com/dotnet/maui/pull/12104

### CODE REVIEW

File: src/Core/src/Core.csproj [hunk 1]
```diff
@@ -21,7 +21,6 @@
   <ItemGroup Condition="$(TargetFramework.Contains('-windows'))">
     <PackageReference Include="Microsoft.WindowsAppSDK" />
     <PackageReference Include="Microsoft.Graphics.Win2D" />
-    <ProjectReference Include="..\..\Graphics\src\Graphics.Win2D\Graphics.Win2D.csproj" />
```
Comment (mattleibow) [inline src/Core/src/Core.csproj:24]: Maui no longer uses the Win2D version, so this is a breaking change as we now use new types and don't automatically bring this package in.

File: src/Core/src/Handlers/ShapeView/ShapeViewHandler.Windows.cs [hunk 2]
```diff
@@ -1,11 +1,11 @@
-﻿using Microsoft.Maui.Graphics.Win2D;
+﻿using Microsoft.Maui.Graphics.Platform;
 
 namespace Microsoft.Maui.Handlers
 {
-	public partial class ShapeViewHandler : ViewHandler<IShapeView, W2DGraphicsView>
+	public partial class ShapeViewHandler : ViewHandler<IShapeView, MauiShapeView>
```
Comment (mattleibow) [inline src/Core/src/Handlers/ShapeView/ShapeViewHandler.Windows.cs:5]: This is a breaking change in CORE. Instead of all the platforms using `MauiShapeView` and windows being a bit "special", we all now use this same type: `MauiShapeView`. It is very simple and does not do anything special:

```cs
public class MauiShapeView : PlatformGraphicsView
{
}
```

File: src/Controls/src/Core/PublicAPI/net-windows/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -8,3 +8,11 @@ Microsoft.Maui.Controls.VisualElement.RefreshIsEnabledProperty() -> void
...
+*REMOVED*~override Microsoft.Maui.Controls.Handlers.PolygonHandler.ConnectHandler(Microsoft.Maui.Graphics.Win2D.W2DGraphicsView nativeView) -> void
+*REMOVED*~override Microsoft.Maui.Controls.Handlers.PolygonHandler.DisconnectHandler(Microsoft.Maui.Graphics.Win2D.W2DGraphicsView nativeView) -> void
+*REMOVED*~override Microsoft.Maui.Controls.Handlers.PolylineHandler.ConnectHandler(Microsoft.Maui.Graphics.Win2D.W2DGraphicsView nativeView) -> void
+*REMOVED*~override Microsoft.Maui.Controls.Handlers.PolylineHandler.DisconnectHandler(Microsoft.Maui.Graphics.Win2D.W2DGraphicsView nativeView) -> void
+~override Microsoft.Maui.Controls.Handlers.PolygonHandler.ConnectHandler(Microsoft.Maui.Platform.MauiShapeView nativeView) -> void
+~override Microsoft.Maui.Controls.Handlers.PolygonHandler.DisconnectHandler(Microsoft.Maui.Platform.MauiShapeView nativeView) -> void
+~override Microsoft.Maui.Controls.Handlers.PolylineHandler.ConnectHandler(Microsoft.Maui.Platform.MauiShapeView nativeView) -> void
+~override Microsoft.Maui.Controls.Handlers.PolylineHandler.DisconnectHandler(Microsoft.Maui.Platform.MauiShapeView nativeView) -> void
```
Comment (mattleibow) [inline src/Controls/src/Core/PublicAPI/net-windows/PublicAPI.Unshipped.txt:18]: The breaking changes.

File: src/Core/src/PublicAPI/net-windows/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -1,2 +1,14 @@
...
+static Microsoft.Maui.Platform.ShapesExtensions.InvalidateShape(this Microsoft.Maui.Graphics.Platform.PlatformGraphicsView! platformView, Microsoft.Maui.IShapeView! shapeView) -> void
+static Microsoft.Maui.Platform.ShapesExtensions.UpdateShape(this Microsoft.Maui.Graphics.Platform.PlatformGraphicsView! platformView, Microsoft.Maui.IShapeView! shapeView) -> void
+*REMOVED*Microsoft.Maui.Handlers.IShapeViewHandler.PlatformView.get -> Microsoft.Maui.Graphics.Win2D.W2DGraphicsView!
+*REMOVED*override Microsoft.Maui.Handlers.ShapeViewHandler.CreatePlatformView() -> Microsoft.Maui.Graphics.Win2D.W2DGraphicsView!
+*REMOVED*static Microsoft.Maui.Platform.GraphicsViewExtensions.UpdateDrawable(this Microsoft.Maui.Graphics.Win2D.W2DGraphicsView! PlatformGraphicsView, Microsoft.Maui.IGraphicsView! graphicsView) -> void
+*REMOVED*static Microsoft.Maui.Platform.ShapesExtensions.InvalidateShape(this Microsoft.Maui.Graphics.Win2D.W2DGraphicsView! platformView, Microsoft.Maui.IShapeView! shapeView) -> void
+*REMOVED*static Microsoft.Maui.Platform.ShapesExtensions.UpdateShape(this Microsoft.Maui.Graphics.Win2D.W2DGraphicsView! platformView, Microsoft.Maui.IShapeView! shapeView) -> void
```
Comment (mattleibow) [inline src/Core/src/PublicAPI/net-windows/PublicAPI.Unshipped.txt:14]: More breaking changes.

File: src/Graphics/src/Graphics/Platforms/Windows/PlatformCanvas.cs [hunk 5]
```diff
@@ -13,9 +13,19 @@
 using WColors = global::Microsoft.UI.Colors;
 #endif
 
+#if MAUI_GRAPHICS_WIN2D
 namespace Microsoft.Maui.Graphics.Win2D
+#else
+namespace Microsoft.Maui.Graphics.Platform
+#endif
```
Comment (mattleibow) [inline src/Graphics/src/Graphics/Platforms/Windows/PlatformCanvas.cs:20]: Same code file, different namespaces.

File: src/Graphics/src/Graphics/Platforms/Windows/PlatformCanvas.cs [hunk 6]
```diff
@@ -13,9 +13,19 @@
...
 {
-	public class W2DCanvas : AbstractCanvas<W2DCanvasState>, IBlurrableCanvas
+#if MAUI_GRAPHICS_WIN2D
+	[System.Obsolete("Use Microsoft.Maui.Graphics.Platform.PlatformCanvas instead.")]
+	public class W2DCanvas
+#else
+	public class PlatformCanvas
+#endif
+		: AbstractCanvas<PlatformCanvasState>, IBlurrableCanvas
```
Comment (mattleibow) [inline src/Graphics/src/Graphics/Platforms/Windows/PlatformCanvas.cs:28]: Same code file, different types. Old type is obsolete, new type is named correctly.

File: src/Graphics/src/Graphics/Platforms/Windows/PlatformCanvas.cs [hunk 7]
```diff
@@ -41,8 +51,12 @@ public class W2DCanvas : AbstractCanvas<W2DCanvasState>, IBlurrableCanvas
 
 		private bool _bitmapPatternFills;
 
+#if MAUI_GRAPHICS_WIN2D
 		public W2DCanvas()
-			: base(new W2DCanvasStateService(), new W2DStringSizeService())
+#else
+		public PlatformCanvas()
+#endif
+			: base(new PlatformCanvasStateService(), new PlatformStringSizeService())
```
Comment (mattleibow) [inline src/Graphics/src/Graphics/Platforms/Windows/PlatformCanvas.cs:59]: Same code file, different constructor type.

**Review by jfversluis — APPROVED** [review]

**Comment by mattleibow** [comment]
Scary part of this is that this suddenly breaks... I am strongly considering doing this in a new way and trying to obsolete but still use or somehow. Maybe we switch in net9 so that people have a chance...

**Comment by techfan101** [comment]
Just curious, is this getting any closer to release? I first became encountered the issue almost a year ago in June of 2022, back in .NET 6...

[...'Platform' does not exist...](https://github.com/dotnet/Microsoft.Maui.Graphics/issues/457)

I've followed it ever since, through its opening/closing as multiple different issues/pulls, continuing to hope that at some point images will be truly supported in a cross-platform manner.

Images seem like a fairly basic UI requirement to me, especially for a cross-platform framework like MAUI. I'm surprised a fix is taking this long...it missed the .NET 7 release. I'm starting to worry that it's going to miss .NET 8 release too. I'd hate to see this same discussion continuing when .NET 9 rolls around.

Thanks, in advance, for any updated information.

**Comment by mattleibow** [comment]
@techfan101 I merged some parts of this PR in https://github.com/dotnet/maui/pull/14816

I am hoping to get the final form into .NET 8 - the main issue is that anything we do may be a breaking change, so we are going slowly and we had to first implement several features in MAUI to aid migration.

**Comment by mattleibow** [comment]
@techfan101 In the next .NET 8 preview you should be able to use PlatformImage and things like that. Just not all the handlers will use the new view types. But this is platform specific code so it does not really make a difference.

TL;DR: should be ready in the next preview - 5 I think.

**Review by jsuarezruiz — COMMENTED** [review]
Require some small changes in WrapperView, ContentView classes to pass the Build.

```
D:\a\_work\1\s\src\Core\src\Platform\Windows\WrapperView.cs(95,28): error CS1061: 'PathF' does not contain a definition for 'AsPath' and no accessible extension method 'AsPath' accepting a first argument of type 'PathF' could be found (are you missing a using directive or an assembly reference?) [D:\a\_work\1\s\src\Core\src\Core.csproj::TargetFramework=net7.0-windows10.0.19041]
D:\a\_work\1\s\src\Core\src\Platform\Windows\ContentPanel.cs(152,28): error CS1061: 'PathF' does not contain a definition for 'AsPath' and no accessible extension method 'AsPath' accepting a first argument of type 'PathF' could be found (are you missing a using directive or an assembly reference?) [D:\a\_work\1\s\src\Core\src\Core.csproj::TargetFramework=net7.0-windows10.0.19041]
```

**Comment by mattleibow** [comment]
Closing this PR for now for some reasons:

 - this will break people unnecessarily
 - we can always reopen for net9
 - the original issue of missing types has been fixed

**Comment by DaveNatalieTripArc** [comment]
Can you explain why this was closed? Will we be able to use the Graphics APIs in Windows Maui app in .Net 8?

**Comment by techfan101** [comment]
I'm very disappointed to see this issue closed. It has been languishing for over a year now (in one form or another). It seems to have moved only slightly closer to resolution.

I've stuck with the Microsoft development stack now for quite a long-time (decades).  Recently, I've been hoping that MAUI was going to keep it relevant. Increasingly, this seems not to be the case.

The pace at which critical issues (and this is only one) are being resolved seems to ensure MAUI's near-term obsolescence. I think this is the year I'll very reluctantly make a clean break and choose not to renew my VS Pro subscription. The benefits no longer justify the expense.

I truly hope more resources are dedicated to this important project and MAUI enjoys some form of future revival. Best of luck.

**Comment by mattleibow** [comment]
> Can you explain more why this was closed? Will we be able to use the Graphics APIs in Windows Maui app in .Net 8?

This was closed because the changes in this PR does not really improve much except to make platform names the same.

We used to use the W2DGraphicsView control as the base of GraphicsView and ShapeView, but this PR would use the new MauiShapeView. This makes me feel good to see all on the same name, but will break everyone.

All the other missing/broken types were fixed other PRs already. As of today, there are 2 native graphics views for windows W2DGraphicsView AND PlatformGraphicsView. You can use either one in your code. And both work with things like PlatformImage.

**Comment by techfan101** [comment]
@mattleibow thank you for the clarification. While I would prefer to see a somewhat cleaner resolution, this does appear to address my core concerns.

I have deleted my prior post, for which I apologize. It was hasty and based on a misunderstanding of the current state of the issue.

It has been quite difficult to follow its status across all of the interrelated postings. Your concise explanation of the (current) resolution helped me considerably.

That said, I do remain concerned about the pace at which issues like this are resolved. I appreciate that the team working on it is working very hard and doing a great job. However, the overall effort seems (perhaps) understaffed at times.
